simple/multi_mr.c

```diff
@@ -0,0 +1,496 @@
+/*
+ * Copyright (c) 2013-2017 Intel Corporation.  All rights reserved.
+ *
+ * This software is available to you under the BSD license
+ * below:
+ *
+ *     Redistribution and use in source and binary forms, with or
+ *     without modification, are permitted provided that the following
+ *     conditions are met:
+ *
+ *      - Redistributions of source code must retain the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer.
+ *
+ *      - Redistributions in binary form must reproduce the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer in the documentation and/or other materials
+ *        provided with the distribution.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AWV
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <getopt.h>
+
+#include <rdma/fabric.h>
+#include <rdma/fi_errno.h>
+#include <rdma/fi_endpoint.h>
+#include <rdma/fi_rma.h>
+#include <rdma/fi_cm.h>
+
+#include "shared.h"
+
+struct test_mr_state {
+	uint8_t *buf;
+	struct fid_cntr *rcntr;
+	uint64_t rcntr_next;
+	struct fid_mr *mr;
+	struct fi_rma_iov *remote;
+};
+
+struct test_state {
+	struct test_mr_state *mr_state;
+	uint64_t mr_count;
+};
+
+struct test_state g_test_state;
+
+static int wait_cntr(struct fid_cntr *cntr, uint64_t *cntr_next)
+{
+	/* 30 second wait timeout */
+	int ret = fi_cntr_wait(cntr, *cntr_next, 30 * 1000);
+	if (ret < 0)
+		return ret;
+
+	*cntr_next += 1;
+
+	return 0;
+}
+
+static int free_mr_res()
+{
+	int ret, i;
+
+	if (g_test_state.mr_state) {
+		for (i = 0; i < g_test_state.mr_count; i++) {
+			ret = fi_close(&g_test_state.mr_state[i].mr->fid);
+			if (ret) {
+				FT_PRINTERR("fi_close", ret);
+				return ret;
+			}
+
+			ret = fi_close(&g_test_state.mr_state[i].rcntr->fid);
+			if (ret) {
+				FT_PRINTERR("fi_close", ret);
+				return ret;
+			}
+
+			if (g_test_state.mr_state[i].buf)
+				free(g_test_state.mr_state[i].buf);
+		}
+		free(g_test_state.mr_state);
+	}
+
+	return 0;
+}
+
+static int alloc_multi_mr_res()
+{
+	int ret = 0, i;
+	struct fi_cntr_attr cntr_attr = {0};
+	cntr_attr.events = FI_CNTR_EVENTS_COMP;
+	cntr_attr.wait_obj = FI_WAIT_UNSPEC;
+
+	if ((fi->domain_attr->mr_key_size * 8) < g_test_state.mr_count) {
+		fprintf(stderr, "ERROR: too many memory regions for unique mr keys");
+		return -1;
+	}
+
+	g_test_state.mr_count = 1ULL << g_test_state.mr_count;
+
+	g_test_state.mr_state =
+		malloc(sizeof(struct test_mr_state) * g_test_state.mr_count);
```

Comment:
When allocating, try to use sizeof(some_variable), rather than directly specifying the name of a structure.  A calloc() call may be safer here to ensure all memory is initialized.